File: library/src/main/java/com/fz/network/RetrofitClient.java
package com.fz.network;


import android.text.TextUtils;

import androidx.annotation.NonNull;

import com.fz.gson.GsonFactory;
import com.fz.network.remote.GsonConverterFactory;
import com.fz.okhttp.OkHttpWrapper;

import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.Map;

import okhttp3.CookieJar;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import retrofit2.CallAdapter;
import retrofit2.Converter;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory;

/**
 * https retrofit 构建，解决多域名需要创建多个实例的问题
 *
 * @author dingpeihua
 * @version 1.0
 * @date 2019/11/22 16:31
 */
public class RetrofitClient {
    public static final MediaType MEDIA_TYPE = MediaType.parse("application/json; charset=utf-8");
    private final OkHttpClient mOkHttpClient;
    /**
     * key: service Class
     * value:service instance object
     */
    private final HashMap<Class<?>, Object> services;
    /**
     * key: host  url
     * value retrofit instance object
     */
    private final HashMap<String, Retrofit> retrofits;
    private Converter.Factory mFactory;
    private CallAdapter.Factory mAdapterFactory;
    /**
     * 服务域名地址
     */
    private String mBaseUrl;

    private RetrofitClient(OkHttpClient okHttpClient) {
        this.mOkHttpClient = okHttpClient;
        services = new HashMap<>();
        retrofits = new HashMap<>();
    }

    private RetrofitClient(Builder builder) {
        this(builder.httpClient);
        this.mBaseUrl = builder.baseUrl;
        this.mFactory = builder.factory;
        this.mAdapterFactory = builder.adapterFactory;
    }

    public void removeRetrofit(String url) {
        retrofits.remove(url);
    }

    public void removeService(Class<?> clazz) {
        services.remove(clazz);
    }

    public void removeAllRetrofit() {
        retrofits.clear();
    }

    public void removeAllService() {
        services.clear();
    }

    public CookieJar getCookieJar() {
        return mOkHttpClient != null ? mOkHttpClient.cookieJar() : null;
    }

    public RetrofitClient setBaseUrl(String mBaseUrl) {
        this.mBaseUrl = mBaseUrl;
        return this;
    }

    /**
     * 创建服务端请求对象
     *
     * @param host
     * @param clazz
     * @author dingpeihua
     * @date 2016/12/23 09:40
     * @version 1.0
     */
    public <T> T createRetrofit(String host, Class<T> clazz, MediaType mediaType, final Type type, final Object typeAdapter) {
        return createRetrofit(host, mOkHttpClient, clazz, mediaType, new HashMap<Type, Object>() {{
            if (type != null && typeAdapter != null) {
                put(type, typeAdapter);
            }
        }});
    }

    private <T> T createRetrofit(String host, OkHttpClient okHttpClient, Class<T> clazz, MediaType mediaType,
                                 Map<Type, Object> typeAdapters,
                                 Converter.Factory factory,
                                 CallAdapter.Factory adapterFactory) {
        if (services.containsKey(clazz)) {
            return clazz.cast(services.get(clazz));
        }
        Retrofit retrofit;
        if (retrofits.containsKey(host)) {
            retrofit = retrofits.get(host);
        } else {
            retrofit = createRetrofit(host, okHttpClient, mediaType, typeAdapters, factory, adapterFactory);
            retrofits.put(host, retrofit);
        }
        T service = retrofit.create(clazz);
        services.put(clazz, service);
        return service;
    }

    /**
     * 创建服务端请求对象
     *
     * @param host
     * @param clazz
     * @author dingpeihua
     * @date 2016/12/23 09:40
     * @version 1.0
     */
    public <T> T createRetrofit(String host, OkHttpClient okHttpClient, Class<T> clazz, MediaType mediaType,
                                Map<Type, Object> typeAdapters) {
        if (services.containsKey(clazz)) {
            return clazz.cast(services.get(clazz));
        }
        Retrofit retrofit;
        if (retrofits.containsKey(host)) {
            retrofit = retrofits.get(host);
        } else {
            retrofit = createRetrofit(host, okHttpClient, mediaType, typeAdapters, null, null);
            retrofits.put(host, retrofit);
        }
        T service = retrofit.create(clazz);
        services.put(clazz, service);
        return service;
    }

    /**
     * 创建一个Retrofit
     *
     * @param host
     * @author dingpeihua
     * @date 2019/1/11 15:26
     * @version 1.0
     */
    private Retrofit createRetrofit(String host, MediaType mediaType, final Type type, final Object typeAdapter,
                                    Converter.Factory factory,
                                    CallAdapter.Factory adapterFactory) {
        return createRetrofit(host, mOkHttpClient, mediaType, new HashMap<Type, Object>() {{
            if (type != null && typeAdapter != null) {
                put(type, typeAdapter);
            }
        }}, factory, adapterFactory);
    }

    /**
     * 创建一个Retrofit
     *
     * @param host
     * @author dingpeihua
     * @date 2019/1/11 15:26
     * @version 1.0
     */
    private Retrofit createRetrofit(String host, OkHttpClient okHttpClient, MediaType mediaType, Map<Type, Object> typeAdapters,
                                    Converter.Factory factory,
                                    CallAdapter.Factory adapterFactory) {
        if (TextUtils.isEmpty(host)) {
            host = mBaseUrl;
        }
        if (factory == null) {
            factory = this.mFactory != null ? this.mFactory : GsonConverterFactory.create(GsonFactory.createBuild(typeAdapters), mediaType);
        }
        if (adapterFactory == null) {
            adapterFactory = this.mAdapterFactory != null ? mAdapterFactory : RxJava3CallAdapterFactory.create();
        }
        return createRetrofit(host, okHttpClient, factory, adapterFactory);
    }

    /**
     * 创建一个Retrofit
     *
     * @param host
     * @author dingpeihua
     * @date 2019/1/11 15:26
     * @version 1.0
     */
    private Retrofit createRetrofit(String host, OkHttpClient okHttpClient,
                                    @NonNull Converter.Factory factory,
                                    @NonNull CallAdapter.Factory adapterFactory) {
        if (TextUtils.isEmpty(host)) {
            host = mBaseUrl;
        }
        return new Retrofit.Builder()
                .baseUrl(host)
                .client(okHttpClient == null ? this.mOkHttpClient : okHttpClient)
                .addConverterFactory(factory)
                .addCallAdapterFactory(adapterFactory)
                .build();
    }

    /**
     * 创建 retorfit
     *
     * @param clazz
     * @param <T>
     * @return
     */
    public <T> T createRetrofit(Class<T> clazz) {
        return createRetrofit(mBaseUrl, clazz);
    }

    /**
     * 创建 retorfit
     *
     * @param clazz
     * @param <T>
     * @return
     */
    public <T> T createRetrofit(String host, Class<T> clazz, MediaType mediaType) {
        return createRetrofit(host, clazz, mediaType, null, null);
    }

    /**
     * 创建 retorfit
     *
     * @param clazz
     * @param <T>
     * @return
     */
    public <T> T createRetrofit(String host, Class<T> clazz, Type type, Object typeAdapter) {
        return createRetrofit(host, clazz, MEDIA_TYPE, type, typeAdapter);
    }

    /**
     * 创建 retorfit
     *
     * @param clazz
     * @param <T>
     * @return
     */
    public <T> T createRetrofit(String host, Class<T> clazz) {
        return createRetrofit(host, clazz, MEDIA_TYPE);
    }

    /**
     * 创建API 服务代理对象
     *
     * @param client
     * @author dingpeihua
     * @date 2019/6/13 11:22
     * @version 1.0
     */
    public static <T> ServiceBuilder<T> createService(RetrofitClient client, Class<T> val) {
        return new ServiceBuilder<>(client, val);
    }

    /**
     * Api 服务代理创建者
     *
     * @author dingpeihua
     * @version 1.0
     * @date 2019/6/13 11:22
     */
    public static final class ServiceBuilder<T> {
        private final RetrofitClient httpClient;
        private OkHttpClient okHttpClient;
        private String host;
        private MediaType mediaType;
        private Class<T> service;
        private Converter.Factory converterFactory;
        private CallAdapter.Factory adapterFactory;
        private HashMap<Type, Object> typeAdapters = new HashMap<>();

        public ServiceBuilder(RetrofitClient vpHttpClient, Class<T> val) {
            this.httpClient = vpHttpClient;
            service = val;
        }

        public ServiceBuilder<T> setHttpClient(OkHttpClient val) {
            okHttpClient = val;
            return this;
        }

        public ServiceBuilder<T> converter(Converter.Factory factory) {
            converterFactory = factory;
            return this;
        }

        public ServiceBuilder<T> adapter(CallAdapter.Factory factory) {
            adapterFactory = factory;
            return this;
        }

        public ServiceBuilder<T> setHttpClient(OkHttpWrapper val) {
            okHttpClient = val.build();
            return this;
        }

        public ServiceBuilder<T> setHost(String val) {
            host = val;
            return this;
        }

        public ServiceBuilder<T> setMediaType(MediaType val) {
            mediaType = val;
            return this;
        }

        public ServiceBuilder<T> setService(Class<T> val) {
            service = val;
            return this;
        }

        public ServiceBuilder<T> setTypeAdapter(Type val1, Object val2) {
            typeAdapters.put(val1, val2);
            return this;
        }

        public T build() {
            return service.cast(httpClient.createRetrofit(host, okHttpClient, service, mediaType, typeAdapters,
                    converterFactory, adapterFactory));
        }
    }

    public static Builder newBuilder() {
        return new Builder();
    }

    public static class Builder {
        private OkHttpClient httpClient;
        private String baseUrl;
        private Converter.Factory factory;
        private CallAdapter.Factory adapterFactory;

        public Builder() {
        }

        public Builder converter(Converter.Factory factory) {
            this.factory = factory;
            return this;
        }

        public Builder adapter(CallAdapter.Factory factory) {
            this.adapterFactory = factory;
            return this;
        }

        public Builder setHttpClient(OkHttpClient httpClient) {
            this.httpClient = httpClient;
            return this;
        }

        public Builder setHttpClient(OkHttpWrapper httpClient) {
            this.httpClient = httpClient.build();
            return this;
        }

        public Builder setBaseUrl(String baseUrl) {
            this.baseUrl = baseUrl;
            return this;
        }

        public RetrofitClient build() {
            return new RetrofitClient(this);
        }
    }
}

File: library/src/main/java/com/fz/network/RetrofitClient.kt
package com.fz.network

import com.fz.gson.GsonFactory
import com.fz.network.remote.GsonConverterFactory.Companion.create
import com.fz.okhttp.OkHttpWrapper
import okhttp3.CookieJar
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import retrofit2.CallAdapter
import retrofit2.Converter
import retrofit2.Retrofit
import java.lang.NullPointerException
import java.lang.reflect.Type
import java.util.*

/**
 * https retrofit 构建，解决多域名需要创建多个实例的问题
 *
 * @author dingpeihua
 * @version 1.0
 * @date 2019/11/22 16:31
 */
class RetrofitClient private constructor(
    private val mOkHttpClient: OkHttpClient,
    private var mAdapterFactory: CallAdapter.Factory? = null,
    private var mFactory: Converter.Factory? = null
) {
    /**
     * key: service Class
     * value:service instance object
     */
    private val services: HashMap<Class<*>, Any> = HashMap()

    /**
     * key: host  url
     * value retrofit instance object
     */
    private val retrofits: HashMap<String, Retrofit> = HashMap()


    /**
     * 服务域名地址
     */
    private var mBaseUrl: String = ""

    private constructor(builder: Builder) : this(builder.httpClient!!) {
        mBaseUrl = builder.baseUrl
        mFactory = builder.factory
        mAdapterFactory = builder.adapterFactory
    }

    fun removeRetrofit(url: String?) {
        retrofits.remove(url)
    }

    fun setAdapterFactory(adapterFactory: CallAdapter.Factory) {
        this.mAdapterFactory = adapterFactory
    }

    fun removeService(clazz: Class<*>) {
        services.remove(clazz)
    }

    fun removeAllRetrofit() {
        retrofits.clear()
    }

    fun removeAllService() {
        services.clear()
    }

    val cookieJar: CookieJar
        get() = mOkHttpClient.cookieJar

    fun setBaseUrl(mBaseUrl: String): RetrofitClient {
        this.mBaseUrl = mBaseUrl
        return this
    }

    /**
     * 创建服务端请求对象
     *
     * @param host
     * @param clazz
     * @author dingpeihua
     * @date 2016/12/23 09:40
     * @version 1.0
     */
    fun <T> createRetrofit(
        host: String,
        clazz: Class<T>,
        mediaType: MediaType?,
        type: Type?,
        typeAdapter: Any?
    ): T {
        return createRetrofit(host, mOkHttpClient, clazz, mediaType, object : HashMap<Type, Any>() {
            init {
                if (type != null && typeAdapter != null) {
                    put(type, typeAdapter)
                }
            }
        })
    }

    private fun <T> createRetrofit(
        host: String, okHttpClient: OkHttpClient?, clazz: Class<T>, mediaType: MediaType?,
        typeAdapters: Map<Type, Any>,
        factory: Converter.Factory?,
        adapterFactory: CallAdapter.Factory?,
    ): T {
        if (services.containsKey(clazz)) {
            return clazz.cast(services[clazz])
        }
        val retrofit: Retrofit = if (retrofits.containsKey(host)) {
            retrofits[host]!!
        } else {
            createRetrofit(host, okHttpClient, mediaType, typeAdapters, factory, adapterFactory)
        }
        retrofits[host] = retrofit
        val service: T = retrofit.create(clazz)
        services[clazz] = service!!
        return service
    }

    /**
     * 创建服务端请求对象
     *
     * @param host
     * @param clazz
     * @author dingpeihua
     * @date 2016/12/23 09:40
     * @version 1.0
     */
    fun <T> createRetrofit(
        host: String, okHttpClient: OkHttpClient?, clazz: Class<T>, mediaType: MediaType?,
        typeAdapters: Map<Type, Any>?,
    ): T {
        if (services.containsKey(clazz)) {
            return clazz.cast(services[clazz])
        }
        val retrofit: Retrofit = if (retrofits.containsKey(host)) {
            retrofits[host]!!
        } else {
            createRetrofit(host, okHttpClient, mediaType, typeAdapters, null, null)
        }
        retrofits[host] = retrofit
        val service = retrofit.create(clazz)
        services[clazz] = service!!
        return service
    }

    /**
     * 创建一个Retrofit
     *
     * @param host
     * @author dingpeihua
     * @date 2019/1/11 15:26
     * @version 1.0
     */
    private fun createRetrofit(
        host: String, mediaType: MediaType, type: Type?, typeAdapter: Any?,
        factory: Converter.Factory,
        adapterFactory: CallAdapter.Factory,
    ): Retrofit {
        return createRetrofit(host, mOkHttpClient, mediaType, object : HashMap<Type, Any>() {
            init {
                if (type != null && typeAdapter != null) {
                    put(type, typeAdapter)
                }
            }
        }, factory, adapterFactory)
    }

    /**
     * 创建一个Retrofit
     *
     * @param host
     * @author dingpeihua
     * @date 2019/1/11 15:26
     * @version 1.0
     */
    private fun createRetrofit(
        host: String?,
        okHttpClient: OkHttpClient?,
        mediaType: MediaType?,
        typeAdapters: Map<Type, Any>?,
        factory: Converter.Factory?,
        adapterFactory: CallAdapter.Factory?,
    ): Retrofit {
        val factory1 = factory ?: mFactory
        ?: create(GsonFactory.createBuild(typeAdapters), mediaType ?: MEDIA_TYPE)
        val adapterFactory1 = adapterFactory ?: mAdapterFactory
        return createRetrofit(host, okHttpClient, factory1, adapterFactory1)
    }

    /**
     * 创建一个Retrofit
     *
     * @param host
     * @author dingpeihua
     * @date 2019/1/11 15:26
     * @version 1.0
     */
    private fun createRetrofit(
        host: String?, okHttpClient: OkHttpClient?,
        factory: Converter.Factory?,
        adapterFactory: CallAdapter.Factory?,
    ): Retrofit {
        val builder = Retrofit.Builder()
            .baseUrl(host ?: mBaseUrl)
            .client(okHttpClient ?: mOkHttpClient)
        if (factory != null) {
            builder.addConverterFactory(factory)
        }
        if (adapterFactory != null) {
            builder.addCallAdapterFactory(adapterFactory)
        }
        return builder.build()
    }

    /**
     * 创建 retorfit
     *
     * @param clazz
     * @param <T>
     * @return
    </T> */
    fun <T> createRetrofit(clazz: Class<T>): T {
        return createRetrofit(mBaseUrl, clazz)
    }

    /**
     * 创建 retorfit
     *
     * @param clazz
     * @param <T>
     * @return
    </T> */
    fun <T> createRetrofit(host: String, clazz: Class<T>, mediaType: MediaType?): T {
        return createRetrofit(host, clazz, mediaType, null, null)
    }

    /**
     * 创建 retorfit
     *
     * @param clazz
     * @param <T>
     * @return
    </T> */
    fun <T> createRetrofit(host: String, clazz: Class<T>, type: Type?, typeAdapter: Any?): T {
        return createRetrofit(host, clazz, MEDIA_TYPE, type, typeAdapter)
    }

    /**
     * 创建 retorfit
     *
     * @param clazz
     * @param <T>
     * @return
    </T> */
    fun <T> createRetrofit(host: String, clazz: Class<T>): T {
        return createRetrofit(host, clazz, MEDIA_TYPE)
    }

    /**
     * Api 服务代理创建者
     *
     * @author dingpeihua
     * @version 1.0
     * @date 2019/6/13 11:22
     */
    class ServiceBuilder<T>(private val httpClient: RetrofitClient, private var service: Class<T>) {
        private var okHttpClient: OkHttpClient? = null
        private var host: String = ""
        private var mediaType: MediaType? = null
        private var converterFactory: Converter.Factory? = null
        private var adapterFactory: CallAdapter.Factory? = null
        private val typeAdapters = HashMap<Type, Any>()
        fun setHttpClient(`val`: OkHttpClient?): ServiceBuilder<T> {
            okHttpClient = `val`
            return this
        }

        fun converter(factory: Converter.Factory?): ServiceBuilder<T> {
            converterFactory = factory
            return this
        }

        fun adapter(factory: CallAdapter.Factory?): ServiceBuilder<T> {
            adapterFactory = factory
            return this
        }

        fun setHttpClient(`val`: OkHttpWrapper): ServiceBuilder<T> {
            okHttpClient = `val`.build()
            return this
        }

        fun setHost(`val`: String): ServiceBuilder<T> {
            host = `val`
            return this
        }

        fun setMediaType(`val`: MediaType?): ServiceBuilder<T> {
            mediaType = `val`
            return this
        }

        fun setService(`val`: Class<T>): ServiceBuilder<T> {
            service = `val`
            return this
        }

        fun setTypeAdapter(val1: Type, val2: Any): ServiceBuilder<T> {
            typeAdapters[val1] = val2
            return this
        }

        fun build(): T {
            return httpClient.createRetrofit(
                host, okHttpClient, service, mediaType, typeAdapters,
                converterFactory, adapterFactory
            )
        }
    }

    class Builder {
        var httpClient: OkHttpClient? = null
        var baseUrl: String = ""
        var factory: Converter.Factory? = null
        var adapterFactory: CallAdapter.Factory? = null
        fun converter(factory: Converter.Factory?): Builder {
            this.factory = factory
            return this
        }

        fun adapter(factory: CallAdapter.Factory?): Builder {
            adapterFactory = factory
            return this
        }

        fun setHttpClient(httpClient: OkHttpClient): Builder {
            this.httpClient = httpClient
            return this
        }

        fun setHttpClient(httpClient: OkHttpWrapper): Builder {
            this.httpClient = httpClient.build()
            return this
        }

        fun setBaseUrl(baseUrl: String): Builder {
            this.baseUrl = baseUrl
            return this
        }

        fun build(): RetrofitClient {
            return RetrofitClient(this)
        }
    }

    companion object {
        val MEDIA_TYPE: MediaType = "application/json; charset=utf-8".toMediaType()

        /**
         * 创建API 服务代理对象
         *
         * @param client
         * @author dingpeihua
         * @date 2019/6/13 11:22
         * @version 1.0
         */
        fun <T> createService(client: RetrofitClient, `val`: Class<T>): ServiceBuilder<T> {
            return ServiceBuilder(client, `val`)
        }

        fun newBuilder(): Builder {
            return Builder()
        }
    }

}